src/hpke.rs

```diff
@@ -0,0 +1,1046 @@
+                    self.aead,
+                    &key_out[..self.aead.key_len()],
+                    &HpkeNonce {
+                        base_nonce: nonce_out,
+                        nonce_counter: [0u8; aead::NONCE_LEN],
+                    },
+                ),
+                opening_key: make_key(
+                    self.aead,
+                    &key_out[..self.aead.key_len()],
+                    &HpkeNonce {
+                        base_nonce: nonce_out,
+                        nonce_counter: [0u8; aead::NONCE_LEN],
+                    },
+                ),
+                raw_key: key_out,
+                base_nonce: nonce_out,
+            },
+            exporter_secret,
+            aead_algorithm: self.aead,
+            peer_aead_ctx: None,
+        })
+    }
+
+    /// new_sender_context
+    pub fn new_sender_context(
+        &self,
+        pk_receiver: &[u8],
+        info: &[u8],
+        encapped_key: &mut [u8],
+        rng: Option<&dyn rand::SecureRandom>,
+        seed: Option<&[u8]>,
+    ) -> Result<SenderContext, error::Unspecified> {
+        let mut shared_secret = [0u8; ec::SCALAR_MAX_BYTES];
+        match rng {
+            Some(r) => self.kem.encap(
+                pk_receiver,
+                encapped_key,
+                &mut shared_secret[..self.kem.private_key_length()],
+                r,
+                seed,
+            )?,
+            None => self.kem.encap(
+                pk_receiver,
+                encapped_key,
+                &mut shared_secret[..self.kem.private_key_length()],
+                &rand::SystemRandom::new(),
+                seed,
+            )?,
+        }
+
+        Ok(SenderContext {
+            inner: self.key_schedule(
+                Mode::Base,
+                &shared_secret[..self.kem.private_key_length()],
+                info,
+                None,
+                None,
+            )?,
+        })
+    }
+
+    /// new_receiver_context
+    pub fn new_receiver_context(
+        &self,
+        encapped_key: &[u8],
+        info: &[u8],
+    ) -> Result<ReceiverContext, error::Unspecified> {
+        let mut shared_secret = [0u8; ec::SCALAR_MAX_BYTES];
+        self.kem.decap(
+            encapped_key,
+            &mut shared_secret[..self.kem.private_key_length()],
+        )?;
+        Ok(ReceiverContext {
+            inner: self.key_schedule(
+                Mode::Base,
+                &shared_secret[..self.kem.private_key_length()],
+                info,
+                None,
+                None,
+            )?,
+        })
+    }
+}
+
+impl Kem {
+    /// Create a new KEM from a given KemId
+    pub fn new(kem_id: KemId) -> Result<Self, error::Unspecified> {
+        match kem_id {
+            KemId::P256Sha256 => Ok(Kem {
+                suite_id_kem: get_suite_id_kem(kem_id),
+                kem_id,
+                hash: &hkdf::HKDF_SHA256,
+                agreement: &agreement::ECDH_P256,
```

Comment:
```suggestion
                dh_group: &agreement::ECDH_P256,
```
 agreement is confusing.

Reply:
In the context of Ring, `agreement` makes _more_ sense. `agreement` only supports ephemeral key exchange, for example.